File: packages/testing/src/adapters/registry.ts
/**
 * Adapter Registry
 *
 * Manages registration and lookup of app adapters.
 */

import type { AdapterRegistry, AdapterRegistryOptions, AppAdapter } from './types.js';

// ============================================================================
// Registry Implementation
// ============================================================================

/**
 * Create an adapter registry.
 */
export function createAdapterRegistry(options: AdapterRegistryOptions = {}): AdapterRegistry {
	const adapters = new Map<string, AppAdapter>();
	const { defaultAdapter } = options;

	// Register initial adapters
	for (const adapter of options.adapters ?? []) {
		adapters.set(adapter.name, adapter);
	}

	return {
		register(adapter: AppAdapter): void {
			if (adapters.has(adapter.name)) {
				throw new Error(`Adapter '${adapter.name}' is already registered`);
			}
			adapters.set(adapter.name, adapter);
		},

		get(name: string): AppAdapter | undefined {
			return adapters.get(name);
		},

		list(): AppAdapter[] {
			return Array.from(adapters.values());
		},

		detect(fixture: unknown): AppAdapter | undefined {
			// Try to detect from fixture's 'app' field
			if (isFixtureWithApp(fixture)) {
				const adapter = adapters.get(fixture.app);
				if (adapter) return adapter;
			}

			// Try each adapter's detection logic
			for (const adapter of adapters.values()) {
				if (canAdapterHandleFixture(adapter, fixture)) {
					return adapter;
				}
			}

			// Return default adapter if specified
			if (defaultAdapter) {
				return adapters.get(defaultAdapter);
			}

			return undefined;
		},

		has(name: string): boolean {
			return adapters.has(name);
		},
	};
}

// ============================================================================
// Detection Helpers
// ============================================================================

/**
 * Check if fixture has an 'app' field.
 */
function isFixtureWithApp(fixture: unknown): fixture is { app: string; [key: string]: unknown } {
	return (
		typeof fixture === 'object' &&
		fixture !== null &&
		'app' in fixture &&
		typeof (fixture as Record<string, unknown>).app === 'string'
	);
}

/**
 * Check if an adapter can handle a fixture based on its schema.
 */
function canAdapterHandleFixture(adapter: AppAdapter, fixture: unknown): boolean {
	// If adapter has a validator, use it
	if (adapter.fixture.validate) {
		// We don't want to await here, so this is a sync check
		// For async validation, use detect() externally
		return false;
	}

	// Check for app-specific fields in fixture
	if (isFixtureWithApp(fixture)) {
		return fixture.app === adapter.name;
	}

	return false;
}

// ============================================================================
// Global Registry
// ============================================================================

let globalRegistry: AdapterRegistry | undefined;

/**
 * Get the global adapter registry.
 * Creates one if it doesn't exist.
 */
export function getGlobalRegistry(): AdapterRegistry {
	if (!globalRegistry) {
		globalRegistry = createAdapterRegistry();
	}
	return globalRegistry;
}

/**
 * Set the global adapter registry.
 */
export function setGlobalRegistry(registry: AdapterRegistry): void {
	globalRegistry = registry;
}

/**
 * Reset the global adapter registry.
 */
export function resetGlobalRegistry(): void {
	globalRegistry = undefined;
}

// ============================================================================
// Convenience Functions
// ============================================================================

/**
 * Register an adapter in the global registry.
 */
export function registerAdapter(adapter: AppAdapter): void {
	getGlobalRegistry().register(adapter);
}

/**
 * Get an adapter from the global registry.
 */
export function getAdapter(name: string): AppAdapter | undefined {
	return getGlobalRegistry().get(name);
}

/**
 * List all adapters in the global registry.
 */
export function listAdapters(): AppAdapter[] {
	return getGlobalRegistry().list();
}

/**
 * Detect adapter for a fixture from the global registry.
 */
export function detectAdapter(fixture: unknown): AppAdapter | undefined {
	return getGlobalRegistry().detect(fixture);
}

File: python/src/afd/testing/adapters/registry.py
"""
Adapter registry implementation.

Manages registration and lookup of app adapters.

Port of packages/testing/src/adapters/registry.ts
"""

from __future__ import annotations

from typing import Any

from afd.testing.adapters.types import (
	AdapterRegistryOptions,
	AppAdapter,
)


class _AdapterRegistryImpl:
	"""Concrete adapter registry implementation."""

	def __init__(self, options: AdapterRegistryOptions | None = None) -> None:
		opts = options or AdapterRegistryOptions()
		self._adapters: dict[str, AppAdapter] = {}
		self._default_adapter = opts.default_adapter

		for adapter in opts.adapters or []:
			self._adapters[adapter.name] = adapter

	def register(self, adapter: AppAdapter) -> None:
		if adapter.name in self._adapters:
			raise ValueError(f"Adapter '{adapter.name}' is already registered")
		self._adapters[adapter.name] = adapter

	def get(self, name: str) -> AppAdapter | None:
		return self._adapters.get(name)

	def list(self) -> list[AppAdapter]:
		return list(self._adapters.values())

	def detect(self, fixture: Any) -> AppAdapter | None:
		if _is_fixture_with_app(fixture):
			adapter = self._adapters.get(fixture["app"])
			if adapter:
				return adapter

		for adapter in self._adapters.values():
			if _can_adapter_handle_fixture(adapter, fixture):
				return adapter

		if self._default_adapter:
			return self._adapters.get(self._default_adapter)

		return None

	def has(self, name: str) -> bool:
		return name in self._adapters


def _is_fixture_with_app(fixture: Any) -> bool:
	"""Check if fixture has an 'app' field."""
	return isinstance(fixture, dict) and isinstance(fixture.get("app"), str)


def _can_adapter_handle_fixture(adapter: AppAdapter, fixture: Any) -> bool:
	"""Check if an adapter can handle a fixture based on its schema."""
	if adapter.fixture.validate:
		return False

	if _is_fixture_with_app(fixture):
		return fixture["app"] == adapter.name

	return False


def create_adapter_registry(
	options: AdapterRegistryOptions | None = None,
) -> _AdapterRegistryImpl:
	"""Create an adapter registry."""
	return _AdapterRegistryImpl(options)


# ---------------------------------------------------------------------------
# Global registry
# ---------------------------------------------------------------------------

_global_registry: _AdapterRegistryImpl | None = None


def get_global_registry() -> _AdapterRegistryImpl:
	"""Get the global adapter registry. Creates one if it doesn't exist."""
	global _global_registry
	if _global_registry is None:
		_global_registry = create_adapter_registry()
	return _global_registry


def set_global_registry(registry: _AdapterRegistryImpl) -> None:
	"""Set the global adapter registry."""
	global _global_registry
	_global_registry = registry


def reset_global_registry() -> None:
	"""Reset the global adapter registry."""
	global _global_registry
	_global_registry = None


def register_adapter(adapter: AppAdapter) -> None:
	"""Register an adapter in the global registry."""
	get_global_registry().register(adapter)


def get_adapter(name: str) -> AppAdapter | None:
	"""Get an adapter from the global registry."""
	return get_global_registry().get(name)


def list_adapters() -> list[AppAdapter]:
	"""List all adapters in the global registry."""
	return get_global_registry().list()


def detect_adapter(fixture: Any) -> AppAdapter | None:
	"""Detect adapter for a fixture from the global registry."""
	return get_global_registry().detect(fixture)
